kvdb/postgres/readwrite_tx.go

```diff
@@ -0,0 +1,295 @@
+			return err
+		}
+
+		if !strings.HasPrefix(table, tx.db.prefix+"_") {
+			continue
+		}
+
+		keys = append(keys, []byte(table[len(tx.db.prefix)+1:]))
+	}
+
+	// Fetch binary top level buckets..
+	bucket := newReadWriteBucket(tx, tx.db.hexTableName, nil)
+	err = bucket.ForEach(func(k, v []byte) error {
+		keys = append(keys, k)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Sort keys.
+	sort.Slice(keys, func(i, j int) bool { return bytes.Compare(keys[i], keys[j]) < 0 })
+
+	// Call callback.
+	for _, k := range keys {
+		err = fn(k)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Rollback closes the transaction, discarding changes (if any) if the
+// database was modified by a write transaction.
+func (tx *readWriteTx) Rollback() error {
+	// If the transaction has been closed roolback will fail.
+	if !tx.active {
+		return walletdb.ErrTxClosed
+	}
+
+	err := tx.tx.Rollback()
+	if err != nil {
+		return err
+	}
+	tx.active = false
+
+	if tx.readOnly {
+		tx.db.lock.RUnlock()
+	} else {
+		tx.db.lock.Unlock()
+	}
+
+	return nil
+}
+
+// ReadWriteBucket opens the root bucket for read/write access.  If the
+// bucket described by the key does not exist, nil is returned.
+func (tx *readWriteTx) ReadWriteBucket(key []byte) walletdb.ReadWriteBucket {
+	table := tx.toTableName(key)
+
+	// If the key can't be mapped to a table name, open the bucket from the
+	// hex table.
+	if table == "" {
+		bucket := newReadWriteBucket(tx, tx.db.hexTableName, nil)
+		return bucket.NestedReadWriteBucket(key)
+	}
+
+	var value int
+	err := tx.tx.QueryRowContext(
+		context.TODO(),
+		"SELECT 1 FROM pg_catalog.pg_tables WHERE schemaname='public' and tablename=$1",
+		table,
+	).Scan(&value)
+
+	if err == sql.ErrNoRows {
+		return nil
+	}
+
+	return newReadWriteBucket(tx, table, nil)
+}
+
+func getCreateTableSql(table string) string {
+	return `
+CREATE TABLE IF NOT EXISTS public.` + table + `
+(
+    key bytea NOT NULL,
+    value bytea,
+    parent_id bigint,
+    id bigserial PRIMARY KEY,
+    sequence bigint,
+    CONSTRAINT ` + table + `_parent FOREIGN KEY (parent_id)
+        REFERENCES public.` + table + ` (id) MATCH SIMPLE
+        ON UPDATE NO ACTION
+        ON DELETE CASCADE
+);
+
+CREATE INDEX IF NOT EXISTS ` + table + `_p
+    ON public.` + table + ` USING btree
+    (parent_id ASC NULLS LAST);
+
+CREATE UNIQUE INDEX IF NOT EXISTS ` + table + `_up
+    ON public.` + table + ` USING btree
+    (parent_id ASC NULLS LAST, key ASC NULLS LAST) WHERE parent_id IS NOT NULL;
+
+CREATE UNIQUE INDEX IF NOT EXISTS ` + table + `_unp ON public.` + table + ` (key) WHERE parent_id IS NULL;
+`
+}
+
+// CreateTopLevelBucket creates the top level bucket for a key if it
+// does not exist.  The newly-created bucket it returned.
+func (tx *readWriteTx) CreateTopLevelBucket(key []byte) (walletdb.ReadWriteBucket, error) {
+	table := tx.toTableName(key)
+
+	if table == "" {
+		bucket := newReadWriteBucket(tx, tx.db.hexTableName, nil)
```

Comment:
Could you please add a comment on what we do here?

Reply:
added